syntax = "proto3";

package proto;


import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

// CRUD
message Post {
    int32 post_id = 1;
    string title = 2;
    string description = 3;
    int32 creator_id = 4;
    bool is_private = 5;
    repeated string tags = 6;
    google.protobuf.Timestamp created_at = 7;
    google.protobuf.Timestamp updated_at = 8;
}

message PostRequest {
    int32 post_id = 1;
    int32 creator_id = 2;
}

message PostResponse {
    Post post = 1;
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    int32 creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message DeletePostRequest {
    int32 post_id = 1;
    int32 creator_id = 2;
}

message UpdatePostRequest {
    int32 post_id = 1;
    string title = 2;
    string description = 3;
    bool is_private = 4;
    repeated string tags = 5;
    int32 creator_id = 6;
}

message ListPostsRequest {
    int32 limit = 1;
    int32 offset = 2;
}


message ListPostsResponse {
    repeated Post posts = 1;
}

// Actions
message PostComment {
    int32 post_id = 1;
    int32 comment_owner = 3;
    string comment = 4;
}

message PostLikeRequest {
    int32 post_id = 1;
    int32 caller_id = 2;
}

message PostUnlikeRequest {
    int32 post_id = 1;
    int32 caller_id = 2;
}

message PostCommentRequest {
    int32 post_id = 1;
    int32 caller_id = 2;
    string content = 3;
}

message PostViewRequest {
    int32 post_id = 1;
    int32 caller_id = 3;
}

message PostCommentsListRequest {
    int32 post_id = 1;
}

message PostCommentsListResponse {
    repeated PostComment comments = 1;
}

service PostsService {
    // CRUD
    rpc CreatePost(CreatePostRequest) returns (PostResponse);
    rpc DeletePost(DeletePostRequest) returns (google.protobuf.Empty);
    rpc UpdatePost(UpdatePostRequest) returns (google.protobuf.Empty);
    rpc GetPost(PostRequest) returns (PostResponse);
    rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);

    // Actions
    rpc LikePost(PostLikeRequest) returns (google.protobuf.Empty);
    rpc UnLikePost(PostUnlikeRequest) returns (google.protobuf.Empty);
    rpc CommentPost(PostCommentRequest) returns (google.protobuf.Empty);
    rpc ViewPost(PostViewRequest) returns (google.protobuf.Empty);
    rpc ListCommentsPost(PostCommentsListRequest) returns (PostCommentsListResponse);
}